tools/util/OperationManager.hpp

```diff
@@ -0,0 +1,1564 @@
+//////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2016-19, Lawrence Livermore National Security, LLC and Umpire
+// project contributors. See the COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (MIT)
+//////////////////////////////////////////////////////////////////////////////
+#ifndef REPLAY_OperationManager_HPP
+#define REPLAY_OperationManager_HPP
+
+#include <iostream>
+#include <cstdint>
+#include <vector>
+
+#include "umpire/Allocator.hpp"
+#include "umpire/strategy/AllocationAdvisor.hpp"
+#include "umpire/strategy/SizeLimiter.hpp"
+#include "umpire/strategy/MixedPool.hpp"
+#include "umpire/strategy/MonotonicAllocationStrategy.hpp"
+#include "umpire/strategy/SlotPool.hpp"
+#include "umpire/strategy/ThreadSafeAllocator.hpp"
+#include "umpire/ResourceManager.hpp"
+
+using AllocationFromLog = uint64_t;
+
+class Operation {
+public:
+  using AllocationOpMap = std::unordered_map<AllocationFromLog, Operation*>;
+  std::function<void ()> op;
+
+  Operation(
+      std::vector<umpire::Allocator*>& alloc_array,
+      AllocationOpMap& alloc_operations
+  ) :   m_alloc_array(alloc_array)
+      , m_alloc_operations(alloc_operations)
+  {
+  }
+
+  void run()
+  {
+    op();
+  }
+
+  //
+  // AllocationAdvisor
+  //
+  void bld_advisor(
+      const bool introspection,
+      const std::string& allocator_name,
+      const std::string& base_allocator_name,
+      const std::string& advice_operation,
+      const int device_id
+  )
+  {
+    if (introspection) {
+      op = [=]() {
+        auto& rm = umpire::ResourceManager::getInstance();
+
+        rm.makeAllocator<umpire::strategy::AllocationAdvisor, true>
+          ( allocator_name, rm.getAllocator(base_allocator_name),
+            advice_operation, device_id);
+
+        auto allocator = new umpire::Allocator(rm.getAllocator(allocator_name));
+        this->m_alloc_array.push_back(allocator);
+      };
+    }
+    else {
+      op = [=]() {
+        auto& rm = umpire::ResourceManager::getInstance();
+
+        rm.makeAllocator<umpire::strategy::AllocationAdvisor, false>
+          ( allocator_name, rm.getAllocator(base_allocator_name),
+            advice_operation, device_id);
+
+        auto allocator = new umpire::Allocator(rm.getAllocator(allocator_name));
+        this->m_alloc_array.push_back(allocator);
+      };
+    }
+  }
+
+  void bld_advisor(
```

Comment:
Let's make a single templated function with a parameter pack and forward it to the `rm.makeAllocator` call in the lambda.

Reply:
Yup, that's a good suggestion - @mcfadden8 can you try that?

Reply:
I tried to pack parameters whereever possible.